hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -2267,14 +2270,155 @@ det
   -> Tensor device dtype (Det shape) -- ^ output
 det input = unsafePerformIO $ ATen.cast1 ATen.Managed.det_t input
 
+-- | Type (kind) of dimension indexes that may be positive (forward from 0) or negative (backwards from end)
+data Dim = PDim Nat | NDim Nat
+
+class KnownDim (dim :: Dim) where
+  dimVal :: Int
+
+instance (KnownNat n) => KnownDim (PDim n) where
+  dimVal = natValI @n
+
+instance (KnownNat n) => KnownDim (NDim n) where
+  dimVal = - natValI @n
+
+-- TODO: eliminate or move to 'Torch.Typed.Aux': UnDim, CmpDim, Drop, Take, Last
+-- TODO: maybe generalize 'DimOutOfBound' and use here?
+type family UnDimImpl (dim :: Dim) (last :: Nat) :: Nat where
+  UnDimImpl (PDim dim) _  = dim
+  UnDimImpl (NDim dim) last =
+    If
+      (dim <=? last)
+      (last - dim)
+      (TypeError
+        (Text "Out of bound dimension: -"
+          :<>: ShowType dim
+          :<>: Text " (the tensor is only "
+          :<>: ShowType last
+          :<>: Text "D)"))
+
+type family UnDim (shape :: [Nat]) (dim :: Dim) :: Nat where
+  UnDim shape dim = UnDimImpl dim (ListLength shape)
+
+type family CmpDim (shape :: [Nat]) (dim :: Dim) (dim' :: Dim) where
+  CmpDim shape dim dim' = CmpNat (UnDim shape dim) (UnDim shape dim')
+
+type family Drop (n :: Nat) (xs :: [a]) :: [a] where
+  Drop 0 xs = xs
+  Drop _ '[] = '[]
+  Drop n (x ': xs) = Drop (n - 1) xs
+
+type family Take (n :: Nat) (xs :: [a]) :: [a] where
+  Take 0 xs = '[]
+  Take _ '[] = '[]
+  Take n (x ': xs) = x ': Take (n - 1) xs
+
+type family Last (xs :: [a]) :: a where
+  Last (x ': '[]) = x
+  Last (x ': xs) = Last xs
```

Comment:
@mcwitt awesome, and I'd gladly add these, but I believe those type families already exist elsewhere. check Aux.hs or Tensor.hs

Reply:
@tscholak gotcha, I skimmed `Torch.Typed.Aux` and didn't immediately see these, so left it as a todo to clean this up. I'll definitely give this another pass before marking ready for review.